Source/Dafny/Rewriter.cs

```diff
@@ -970,6 +970,55 @@ public AutoContractsRewriter(ErrorReporter reporter, BuiltIns builtIns)
     }
   }
 
+  public class OpaqueConstRewriter : IRewriter {
+    public OpaqueConstRewriter(ErrorReporter reporter) : base(reporter) {
+    }
+
+    internal override void PreResolve(ModuleDefinition m) {
+      foreach (var d in m.TopLevelDecls) {
+        if (d is TopLevelDeclWithMembers) {
+          ProcessOpaqueClassConsts((TopLevelDeclWithMembers)d);
+        }
+      }
+    }
+
+    private void ProcessOpaqueClassConsts(TopLevelDeclWithMembers c) {
+      Contract.Requires(c != null);
+      List<MemberDecl> newDecls = new List<MemberDecl>();
+      foreach (MemberDecl member in c.Members) {
+        if (member is ConstantField constant) {
+          if (!Attributes.Contains(constant.Attributes, "opaque")) {
```

Comment:
See my comments in `Translator.cs` about having a method `IsOpaque`. I guess such a method can be defined in `DafnyAst.cs` or `Resolver.cs` instead of in `Translator.cs`.

Reply:
I didn't add it yet, but could add it to `MemberDecl`.